content/posts/networkx/vf2pp/graph-iso-vf2pp/index.md

```diff
@@ -0,0 +1,132 @@
+in the two following diagrams.
+
+<center><img src="times.png" alt="vf2++ and vf2 times"/></center>
+<center><img src="speedup.png" alt="speedup"/></center>
+
+We notice that the maximum speedup achieved is **14x**, and it keeps increasing gradually, as the number of nodes increase.
+It is also highly prominent that the increase in number of nodes, doesn't seem to affect the performance of **VF2++** to
+a significant extent, when compared to the drastic impact on the performance of **VF2**. Our results are almost identical
+to those presented in the original **VF2++ paper**, verifying the theoretical analysis and premises of the literature.
+
+## Optimizations
+
+The achieved boost is due to some key improvements and optimizations, and more specifically:
+
+- **Optimal node ordering**, which avoids following unfruitful branches that will result in infeasible states. We make sure that the nodes that have the biggest possibility to match, are accessed first.
+- **Implementation in a non-recursive manner**, which saves both time and space.
+- **Caching** of both node degrees and nodes per degree in the beginning, so that we don't have to access those features in every degree check. For example, instead of doing
+
+```python
+res = []
+for node in G2.nodes():
+    if G1.degree[u] == G2.degree[node]:
+        res.append(node)
+# do stuff with res ...
+```
+
+to get the nodes of same degree as u (which happens a lot of times in the implementation), we just do:
+
+```python
+res = G2_nodes_of_degree[G1.degree[u]]
+# do stuff with res ...
+```
+
+where "G2_nodes_of_degree" stores set of nodes for a given degree. Same happend with node labels.
+
+- **Extra shrinking of the candidate set for each node** by adding more checks in the candidate selection method and removing some from the feasibility checks. In simple terms, instead of checking a lot of conditions on a larger set of candidates, we check fewer conditions but on a more targeted and significantly smaller set of candidates.
+  For example, in this code:
+
+```python
+candidates = set(G2.nodes())
+for candidate in candidates:
+    if feasibility(u, candidate):
+        do_stuff()
+```
+
+we take a huge set of candidates, which results in poor performance due to maximizing calls of "feasibility", thus performing
+the feasibility checks in a very large set. Now compare that to the following alternative:
+
+```python
+candidates = [
+    n
+    for n in G2_nodes_of_degree[G1.degree[u]].intersection(
+        G2_nodes_of_label[G1_labels[u]]
+    )
+]
+for candidate in candidates:
+    if feasibility(u, candidate):
+        do_stuff()
+```
+
+Immediately we have drastically reduced the number of checks performed and calls to the function, as now we only apply them to nodes of the same degree and label as $u$. This is a simplification for demonstration purposes. In the actual implementation there are more checks and extra shrinking of the candidate set.
```

Comment:
This post does a really nice job summarizing the algorithm; IMO, what would really knock this out of the park would be to finish with an example of applying VF2++ to a "classic" isomorphism problem.

It's not *necessary*, but I think this would really be a compelling way to end the series of posts. Can we think of good "canonical" example of an isomorphism problem that we could demo here?

Reply:
@rossbar 
Do you mean providing an example and following the steps of the algorithm one by one, explaining everything? Yup, we could use the graphs included in the isomorphism wikipedia page!

Reply:
I was thinking even lighter-weight than that - no reason to do the step-by-step breakdown, just show the result of vf2pp in action!

Reply:
Ok so for example we can show the mapping returned be the solver or smth